File: java/src/main/java/vsdk/toolkit/numericalAnalysis/quasiMonteCarlo/Halton.java
package vsdk.toolkit.numericalAnalysis.quasiMonteCarlo;

/**
Halton quasi Monte Carlo sample generator
*/
public class Halton {
    public static double Halton2(int i) {
        long h;
        long f;

        h = i & 1L;
        f = 2L;
        i >>= 1;
        while (i != 0) {
            h <<= 1;
            h += (i & 1L);
            i >>= 1;
            f <<= 1;
            h <<= 1;
            h += (i & 1L);
            i >>= 1;
            f <<= 1;
            h <<= 1;
            h += (i & 1L);
            i >>= 1;
            f <<= 1;
            h <<= 1;
            h += (i & 1L);
            i >>= 1;
            f <<= 1;
        }

        return (double)h / (double)f;
    }

    public static double Halton3(int i) {
        long h;
        long f;
        long j = i;
        i /= 3;
        h = j - ((i << 1) + i);
        f = 3;
        while (i > 0) {
            long k;
            j = i;
            i /= 3;
            k = h - i;
            h = j + (k << 1) + k;
            f = (f << 1) + f;
        }

        return (double)h / (double)f;
    }

    public static double Halton5(int i) {
        long h;
        long f;
        long j = i;
        i /= 5;
        h = j - ((i << 2) + i);
        f = 5;
        while (i > 0) {
            long k;
            j = i;
            i /= 5;
            k = h - i;
            h = j + (k << 2) + k;
            f = (f << 2) + f;
        }

        return (double)h / (double)f;
    }

    public static double Halton7(int i) {
        long h;
        long f;
        long j = i;
        i /= 7;
        h = j - ((i << 2) + (i << 1) + i);
        f = 7;
        while (i > 0) {
            long k;
            j = i;
            i /= 7;
            k = h - i;
            h = j + (k << 2) + (k << 1) + k;
            f = (f << 2) + (f << 1) + f;
        }

        return (double)h / (double)f;
    }
}

File: typescript/base/src/main/vsdk/toolkit/numericalAnalysis/quasiMonteCarlo/Halton.ts
export class Halton {
  public static Halton2(i: number): number {
    let value = BigInt(globalThis.Math.trunc(i));
    let h = value & 1n;
    let f = 2n;
    value >>= 1n;
    while (value !== 0n) {
      h <<= 1n;
      h += (value & 1n);
      value >>= 1n;
      f <<= 1n;
      h <<= 1n;
      h += (value & 1n);
      value >>= 1n;
      f <<= 1n;
      h <<= 1n;
      h += (value & 1n);
      value >>= 1n;
      f <<= 1n;
      h <<= 1n;
      h += (value & 1n);
      value >>= 1n;
      f <<= 1n;
    }

    return Number(h) / Number(f);
  }

  public static Halton3(i: number): number {
    let value = BigInt(globalThis.Math.trunc(i));
    let j = value;
    value /= 3n;
    let h = j - ((value << 1n) + value);
    let f = 3n;
    while (value > 0n) {
      const k = h - value;
      j = value;
      value /= 3n;
      h = j + (k << 1n) + k;
      f = (f << 1n) + f;
    }

    return Number(h) / Number(f);
  }

  public static Halton5(i: number): number {
    let value = BigInt(globalThis.Math.trunc(i));
    let j = value;
    value /= 5n;
    let h = j - ((value << 2n) + value);
    let f = 5n;
    while (value > 0n) {
      const k = h - value;
      j = value;
      value /= 5n;
      h = j + (k << 2n) + k;
      f = (f << 2n) + f;
    }

    return Number(h) / Number(f);
  }

  public static Halton7(i: number): number {
    let value = BigInt(globalThis.Math.trunc(i));
    let j = value;
    value /= 7n;
    let h = j - ((value << 2n) + (value << 1n) + value);
    let f = 7n;
    while (value > 0n) {
      const k = h - value;
      j = value;
      value /= 7n;
      h = j + (k << 2n) + (k << 1n) + k;
      f = (f << 2n) + (f << 1n) + f;
    }

    return Number(h) / Number(f);
  }

  private constructor() {
  }
}
